File: windows/DracPaste.Protocolo/Protocolo.cs
namespace DracPaste.Protocolo;

/// <summary>
/// Constantes del protocolo DracPaste v1.
///
/// Todo lo que hay aquí está fijado en <c>docs/protocol.md</c>. Si un valor cambia,
/// cambia primero el documento y después este fichero y su equivalente en Kotlin
/// (<c>Protocolo.kt</c>), en el mismo cambio.
/// </summary>
public static class Protocolo
{
    /// <summary>Versión del protocolo que habla esta implementación.</summary>
    public const int Version = 1;

    /// <summary>Tipo de servicio mDNS que publica el PC y busca el móvil.</summary>
    public const string ServicioMdns = "_dracpaste._tcp";

    /// <summary>Puerto preferido. Si está ocupado, se toma otro y se anuncia por mDNS.</summary>
    public const int PuertoPreferido = 47653;

    /// <summary>Máximo de un frame completo. Un length mayor es error de protocolo.</summary>
    public const int MaxFrameBytes = 1024 * 1024;

    /// <summary>Máximo del texto de un clip antes de codificar.</summary>
    public const int MaxClipBytes = 256 * 1024;

    /// <summary>Cada cuánto se envía un PING.</summary>
    public const int IntervaloPingMs = 15_000;

    /// <summary>Sin PONG en este plazo, la conexión se da por muerta.</summary>
    public const int TimeoutPongMs = 10_000;

    /// <summary>Un handshake que no termina en este plazo se aborta.</summary>
    public const int TimeoutHandshakeMs = 10_000;

    /// <summary>Cuánto vale el token del QR de emparejamiento.</summary>
    public const int ValidezTokenMs = 120_000;

    /// <summary>Ventana en la que un clip recién recibido no se reenvía (anti-eco).</summary>
    public const int VentanaAntiEcoMs = 5_000;

    /// <summary>Varias apps disparan más de un WM_CLIPBOARDUPDATE por una sola copia.</summary>
    public const int DebounceClipboardMs = 100;

    /// <summary>Único tipo de contenido que v1 sabe transportar.</summary>
    public const string TipoTexto = "text/plain";

    /// <summary>Formato de portapapeles propio que marca lo que escribe esta app.</summary>
    public const string FormatoMarcaPropia = "DracPasteOrigin";

    /// <summary>Claves de los registros TXT del anuncio mDNS.</summary>
    public const string TxtVersion = "v";

    /// <inheritdoc cref="TxtVersion"/>
    public const string TxtId = "id";

    /// <inheritdoc cref="TxtVersion"/>
    public const string TxtNombre = "name";
}

File: android/protocolo/src/main/kotlin/com/marcmayol/dracpaste/protocolo/Protocolo.kt
package com.marcmayol.dracpaste.protocolo

/**
 * Constantes del protocolo DracPaste v1.
 *
 * Todo lo que hay aquí está fijado en `docs/protocol.md`. Si un valor cambia, cambia
 * primero el documento y después este fichero y su equivalente en C#, en el mismo cambio.
 */
object Protocolo {
    /** Versión del protocolo que habla esta implementación. */
    const val VERSION = 1

    /** Tipo de servicio mDNS que publica el PC y busca el móvil. */
    const val SERVICIO_MDNS = "_dracpaste._tcp"

    /** Puerto preferido. Si está ocupado, el PC toma otro y lo anuncia por mDNS. */
    const val PUERTO_PREFERIDO = 47653

    /** Máximo de un frame completo. Un `length` mayor se trata como error de protocolo. */
    const val MAX_FRAME_BYTES = 1024 * 1024

    /** Máximo del texto de un clip antes de codificar. */
    const val MAX_CLIP_BYTES = 256 * 1024

    /** Cada cuánto se envía un PING. */
    const val INTERVALO_PING_MS = 15_000L

    /** Sin PONG en este plazo, la conexión se da por muerta. */
    const val TIMEOUT_PONG_MS = 10_000L

    /** Un handshake que no termina en este plazo se aborta. */
    const val TIMEOUT_HANDSHAKE_MS = 10_000L

    /** Espera inicial entre reintentos de conexión. */
    const val BACKOFF_INICIAL_MS = 1_000L

    /** Tope de la espera entre reintentos. */
    const val BACKOFF_MAXIMO_MS = 30_000L

    /** Cuánto vale el token del QR de emparejamiento. */
    const val VALIDEZ_TOKEN_MS = 120_000L

    /** Ventana en la que un clip recién recibido no se reenvía (anti-eco). */
    const val VENTANA_ANTIECO_MS = 5_000L

    /** Cuánto se retiene en memoria un envío que falló por no haber conexión. */
    const val RETENCION_ENVIO_FALLIDO_MS = 60_000L

    /** Único tipo de contenido que v1 sabe transportar. */
    const val TIPO_TEXTO = "text/plain"

    /** Claves de los registros TXT del anuncio mDNS. */
    const val TXT_VERSION = "v"
    const val TXT_ID = "id"
    const val TXT_NOMBRE = "name"
}
